### PR TITLE
Chapter 10: Updating, showing, and deleting users

### PR DESCRIPTION
**Pull request by stephanie1997-kek**

### CODE REVIEW

**Comment by stephanie1997-kek** [comment]
ready

File: Gemfile [hunk 1]
```diff
@@ -7,17 +7,23 @@ end
...
+gem "kaminari"
+gem "kaminari-bootstrap"
 gem "mysql2"
 gem "puma", "~> 3.7"
 gem "rails", "~> 5.1.7"
 gem "rails-i18n"
 gem "sass-rails", "~> 5.0"
 gem "turbolinks", "~> 5"
 gem "uglifier", ">= 1.3.0"
+gem "will_paginate", git: "https://github.com/mislav/will_paginate.git", branch: "rails6"
```
Comment (ngocvt-0484) [inline Gemfile]: if you dont use this gem, please remove it

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -1,5 +1,12 @@
 class UsersController < ApplicationController
-  before_action :load_user, only: :show
+  before_action :logged_in_user, except: %i(show new create)
+  before_action :logged_in_user, except: %i(show new create)
```
Comment (ngocvt-0484) [inline app/controllers/users_controller.rb]: delete this line because you defined it above

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -33,4 +60,21 @@ def load_user
     flash[:warning] = t "users.user_not_found"
     redirect_to root_path
   end
+
+  def logged_in_user
+    unless logged_in?
```
Comment (ngocvt-0484) [inline app/controllers/users_controller.rb]: return unless logged_in?
store_location
flash[:danger] = t "users.login_remind"
redirect_to login_path

File: app/views/users/_form.html.erb [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  <%= f.email_field :email, class: "form-control" %>
+
+  <%= f.label :password %>
+  <%= f.password_field :password, class: "form-control" %>
+
+  <%= f.label :password_confirmation, t(".confirmation") %>
+  <%= f.password_field :password_confirmation, class: "form-control" %>
+
+  <%= f.submit yield(:button_text), class: "btn btn-primary" %>
+<% end %>
```
Comment (ngocvt-0484) [inline app/views/users/_form.html.erb]: please add endline

File: app/views/users/_user.html.erb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+<li>
+  <%= gravatar_for user, size: Settings.collections.gravatar.size.list %>
+  <%= link_to user.name, user %>
+  <% if current_user.admin? && !current_user?(user) %> | <%= link_to t(".delete_link"), user, method: :delete, data: { confirm: t(".confirm") } %>
+  <% end %>
+</li>
```
Comment (ngocvt-0484) [inline app/views/users/_user.html.erb]: please add endline

**Comment by vanvtt-0952** [comment]
commented

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (stephanie1997-kek) [reply app/controllers/users_controller.rb]: I have deleted it before create this pull request.

File: app/views/users/_form.html.erb (hunk 4 above)
Comment (stephanie1997-kek) [reply app/views/users/_form.html.erb]: done

File: app/views/users/_user.html.erb (hunk 5 above)
Comment (stephanie1997-kek) [reply app/views/users/_user.html.erb]: done

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (stephanie1997-kek) [reply app/controllers/users_controller.rb]: done

**Comment by stephanie1997-kek** [comment]
ready

File: app/assets/stylesheets/application.scss [hunk 6]
```diff
@@ -0,0 +1,5 @@
+/*
+ *= require_tree .
+ *= require_self
+ */
+@import 'bootstrap'
```
Comment (chungpv-1008) [inline app/assets/stylesheets/application.scss]: Please add endline for this file

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -33,4 +61,21 @@ def load_user
...
+
+
+  def correct_user
+    redirect_to root_path unless current_user? @user
+  end
+
+  def admin_user
+    redirect_to root_path unless current_user.admin?
+  end
+end
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]: similar

File: app/helpers/users_helper.rb [hunk 8]
```diff
@@ -1,7 +1,8 @@
 module UsersHelper
-  def gravatar_for user
-    gravatar_id = Digest::MD5::hexdigest user.email.downcase
-    gravatar_url = "#{Settings.links.gravatar}#{gravatar_id}"
+  def gravatar_for user, options = {size: Settings.gravartar.size_default}
+    gravatar_id = Digest::MD5::hexdigest (user.email.downcase)
```
Comment (chungpv-1008) [inline app/helpers/users_helper.rb]:
```suggestion
    gravatar_id = Digest::MD5::hexdigest user.email.downcase
```

File: config/settings.yml [hunk 9]
```diff
@@ -21,3 +23,9 @@ validations:
 
 collections:
   session_controller_create: 1
+  gravatar:  
+    size:
+      default: 80
+      list: 50
+gravartar:
+  size_default: 80
```
Comment (chungpv-1008) [inline config/settings.yml]: Please add endline for this file

**Review by chungpv-1008 — COMMENTED** [review]
Please fix comment, then merge this pull request yourself

**Pull request merged**